### DISCUSSION TITLE
Using with Nest.Js

### DISCUSSION THREAD

**Discussion by gax97** [trpc/trpc#1504] [Q&A] [answered]
Hi everyone 👋 

I wonder how suitable this library is for Nest.js?
It would be awesome if we can have an example for nest js projects.

**Comment by sachinraja** [comment]
Nest does not seem like something you would use *with* tRPC. How do you see them working together?

**Reply by carnesen** [reply]
We are considering using tRPC with Nest.js. We like Nest.js's dependency injection but not so much how it organizes HTTP endpoints using @Controller and class-validator/class-transformer for runtime data validation. We like Zod a lot better and how tRPC provides an easier path to type safety.

**Reply by KATT** [reply]
https://github.com/trpc/trpc/discussions/1504#discussioncomment-2176637

**Reply by lkostrowski** [reply]
@carnesen you can use Zod with Nest https://docs.nestjs.com/pipes#object-schema-validation

**Reply by KevinEdry** [reply]
https://www.nestjs-trpc.io/
I've built the integration between the two, its not perfect and i'm still working on the docs and readme, but I'd love to have your feedback on it.

**Comment by nirtamir2** [comment]
I notice #1511 has been resolved.
I wonder if this helps with this task.
It will help Nest.js users to have type-safe api end to end between client and server

**Comment by myesn** [comment]
`tPRC` looks really good and greatly improves development efficiency, but there are too many things to rewrite to adapt existing applications. Is there an easy way to make `tRPC` compatible with existing `NestJS` applications? It would be really awesome if this could be done!

**Comment by reslear** [comment]
same, wait :)

**Comment by reslear** [comment]
Is there any approximate roadmap?

**Comment by KATT** [comment, marked as answer]
None of the maintainers of tRPC are users of nestjs. 

Therefore, we have no plans on putting in more unpaid work to support nestjs when we have no use of it ourselves - this has to come from the community or through paid consulting.

I have no idea what it would take to support nest as I'm not a user of it so I have no insight on how an adapter would look like. 

---

**Update:** [see potential workaround by @hypnodron below](https://github.com/trpc/trpc/discussions/1504#discussioncomment-4218180)

**Reply by judocode** [reply]
Any updates?

**Reply by macstr1k3r** [reply]
@judocode https://github.com/macstr1k3r/trpc-nestjs-adapter

**Reply by KevinEdry** [reply]
https://www.nestjs-trpc.io/
I've been working on this adapter for a while, this combines both the opinionated approach of NestJS (Controllers, Providers and Middlewares) and the type-safety of tRPC.

Here's a brief usage example from the docs:
```ts
// users.router.ts
import { Inject } from '@nestjs/common';
import { Router, Query, Middlewares } from 'nestjs-trpc';
import { UserService } from './user.service';
import { ProtectedMiddleware } from './protected.middleware';
import { TRPCError } from '@trpc/server';
import { z } from 'zod';

const userSchema = z.object({
  name: z.string(),
  password: z.string()
})

@Router()
class UserRouter {
  constructor(
    @Inject(UserService) private readonly userService: UserService
  ) {}

  @Middlewares(ProtectedMiddleware)
  @Query({ output: z.array(userSchema) })
  async getUsers() {
    try {
      return this.userService.getUsers();
    } catch (error: unknown) {
      throw new TRPCError({
        code: "INTERNAL_SERVER_ERROR",
        message: "An error has occured when trying to get users.",
        cause: error
      })
    }
  }
}
```

**Reply by xyunsh** [reply]
Great job

**Reply by afazzdev** [reply]
> https://www.nestjs-trpc.io/ I've been working on this adapter for a while, this combines both the opinionated approach of NestJS (Controllers, Providers and Middlewares) and the type-safety of tRPC.
> 
> Here's a brief usage example from the docs:
> 
> ```ts
> // users.router.ts
> import { Inject } from '@nestjs/common';
> import { Router, Query, Middlewares } from 'nestjs-trpc';
> import { UserService } from './user.service';
> import { ProtectedMiddleware } from './protected.middleware';
> import { TRPCError } from '@trpc/server';
> import { z } from 'zod';
> 
> const userSchema = z.object({
>   name: z.string(),
>   password: z.string()
> })
> 
> @Router()
> class UserRouter {
>   constructor(
>     @Inject(UserService) private readonly userService: UserService
>   ) {}
> 
>   @Middlewares(ProtectedMiddleware)
>   @Query({ output: z.array(userSchema) })
>   async getUsers() {
>     try {
>       return this.userService.getUsers();
>     } catch (error: unknown) {
>       throw new TRPCError({
>         code: "INTERNAL_SERVER_ERROR",
>         message: "An error has occured when trying to get users.",
>         cause: error
>       })
>     }
>   }
> }
> ```

Thx for your work!. It's exactly what i need but i can't make it running within nx project. It always thrown error when i'm adding autoSchemaFile option.  The error is `Error: Could not access root module file.`

**Comment by dheerajbhaskar** [comment]
@KATT said this
> I have no idea what it would take to support nest as I'm not a user of it so I have no insight on how an adapter would look like.

Then it occurred to me, nestjs uses express or fastify engines (as per user choice). Trpc already has adapters for express and fastify. So, I guess we just need to figure out a way to hook into the underlying engine with existing trpc adapters.

Of course, an example project implementation would be helpful, but this can be a starting point

**Comment by jbw** [comment]
I played around with this.  I hadn't used trpc before, but I use NestJS. You _could_ sit alongside the API layer of NestJS  by creating a trpc router like:

```ts
const e = express();
const app = await NestFactory.create<NestExpressApplication>(AppModule, new ExpressAdapter(e));

e.use(
  '/trpc',
  trpcExpress.createExpressMiddleware({
    router: appRouter,
  }),
);

await app.listen(3000);
```

Which completely sidesteps any NestJS controllers, middleware, etc such as validation using `class-validator`. 

But you can use DI and the module system but I imagine there are some caveats when you want to mess with injection scopes. 

For example, getting an existing service and hooking that up to your router. 

```ts
const appService = app.get(AppService);

t.router({
  hello: t.procedure.query(() => {
    return appService.getHello();
  }),
});
```

Before I started, I was thinking it would be nice to keep the NestJS controllers and see if I could pull information like routes, types, and handlers into the trpc router. But you would still lose any middleware, filters, guards, pipes, and interceptors from NestJS (so most of NestJS 😄) . Also, from the trpc docs, it looks like it has its own way of handling those concerns anyway. 

In the end, I don't see much benefit in combing the two.  Maybe find a DI solution for trpc.

**Comment by hypnodron** [comment]
I've recently had some experience with integrating TRPC into a NestJS project. The basic setup looks like this:

trpc.service.ts
```ts
@Injectable()
export class TRPCService {
  constructor(
    // DI provided by NestJS
    @Inject(UserRepo) private readonly userRepo: UserRepo
  ) {}

  // Singleton scope is used for injection by default so `t` is created only once
  t = initTRPC.create()

  appRouter = this.t.router({
    users: this.t.procedure.query(async (opts) => {
      // Here we can use providers injected by NestJS in our procedures 
      return this.userRepo.find()
    })  
  })
  
  applyMiddleware(app: NestExpressApplication) {
    app.use(
      "/trpc",
      createExpressMiddleware({
        router: this.appRouter,
      })
    )
  }
}
```

main.ts
```ts
const app = await NestFactory.create(AppModule)
const trpcService = app.get(TRPCService)
trpcService.applyMiddleware(app)
```

As the project grows `initTRPC` and child routers can be moved into their own providers (injections in constructors skipped for brevity):

trpc-init.service.ts
```ts
@Injectable()
export class TRCPInitService {
  t = initTRPC.create()
} 
```

trpc-users.service.ts
```ts
@Injectable()
export class TRCPUsersService {
  usersRouter = this.trpcInit.t.procedure.query(...)
} 
```

trpc.service.ts
```ts
@Injectable()
export class TRCPService {
  appRouter = this.trpcInit.t.router({
    users: this.trpcUsers.usersRouter
  })
} 
```

**Reply by LoriKarikari** [reply]
How do you import the `AppRouter`  type on the client if you implement it this way?

**Reply by hypnodron** [reply]
```ts
export type AppRouter = TRPCService["appRouter"]
```

**Reply by code-xhyun** [reply]
Can i get the boilerplate?  @hypnodron

**Reply by tomwray13** [reply]
Nice work @hypnodron! I just put the above ideas into a quick [repo](https://github.com/tomwray13/nestjs-trpc) if anyone wants to check out an example using both the trpc server and client in a NestJS monorepo setup.

**Reply by sttvld** [reply]
Thanks for solution, but there is a huge problem - DI.

Most likely every nest app uses a lot request scope services.
`app.resolve` would try to create service instance but request will be empty.

As I understood, Nest creates a DI containers hierarchy for each resolver, so if we try to use `app.resolve` in app level / module scope default(singleton), that DI container doesn't have request.

I see three options in this situation:
- implement every needed service without Nest DI (almost the same if you write a separate application and bundle them with nginx. Session could be shared easily with redis)

- Refactor services and isolate them from directly request usage. So you could provide them manually

- Pass somehow DI resolver with request scope to trpc level. I'm trying this way but no success yet. Any ideas on how to pass resolver?

**Reply by lwhiteley** [reply]
Thanks worked for me.

i also fiddled and added a websocket server to it if anyone is interested in that as well


```ts
// add to TRPCService (or TRPCRouter in repo example)
createWebSocketServer(app: INestApplication, port: string | number) {
    const wss = new ws.Server({
      server: app.getHttpServer(),
    });
    const handler = applyWSSHandler({ wss, router: this.appRouter });

    wss.on('connection', (ws) => {
      Logger.log(`➕➕ Connection (${wss.clients.size})`);
      ws.once('close', () => {
        Logger.log(`➖➖ Connection (${wss.clients.size})`);
      });
    });
   
    process.on('SIGTERM', () => {
      Logger.log('SIGTERM');
      handler.broadcastReconnectNotification();
      wss.close();
      app.close();
    });

    Logger.log(`✅ WebSocket Server listening on ws://localhost:${port}`);
  }
```

```ts
// then my bootstrap function looks like this

async function bootstrap() {
  const app = await NestFactory.create(AppModule);
  const globalPrefix = '';
  app.setGlobalPrefix(globalPrefix);
  const port = process.env.PORT || 3333;
  const trpc = app.get(TrpcRouter);
  trpc.applyMiddleware(app);
  trpc.createWebSocketServer(app, port);
  await app.listen(port);
  Logger.log(
    `🚀 Application is running on: http://localhost:${port}/${globalPrefix}`,
  );
}

```

you can follow the trpc docs from there to setup a client

Has the same DI problem though*

**Reply by nllahat** [reply]
Can u plz share an example of the problem withbthe request scope?
I'm not sure I understand

**Reply by macstr1k3r** [reply]
@nllahat see https://github.com/macstr1k3r/trpc-nestjs-adapter for a solution which works with request scoped providers.

**Reply by LZL0** [reply]
> @nllahat see https://github.com/macstr1k3r/trpc-nestjs-adapter for a solution which works with request scoped providers.

Looks great!

**Reply by AlexSua** [reply]
Maybe there is something that I am missing, but following this example:
```js
@Injectable()
export class TRCPService {
  appRouter = this.trpcInit.t.router({
    users: this.trpcUsers.usersRouter
  })
} 
 ```
 You can't use `this.trpcUsers` there right?  I mean, you are using `this.trpcUsers` that comes from DI before being initialized. And if you put the `appRouter` in the constructor, how could you export to the client like this?:
 
 ```js
 export type AppRouter = TRPCService["appRouter"]
```
 
And this means also you can't use providers inside requests:
 ```js
  appRouter = this.t.router({
    users: this.t.procedure.query(async (opts) => {
      return this.userRepo.find() //<----- this.userRepo not initialized
    })  
  })
 ``` 
 
 And all of this makes me believe that you can't use DI with TRPC... which makes me wonder why would you use Nest with TRPC?

**Reply by hypnodron** [reply]
@AlexSua, the injections in constructors were skipped for brevity, the full example would look something like this: 

```ts
// trpc-init.service.ts
@Injectable()
export class TRCPInitService {
  t = initTRPC.create()
} 
```

```ts
// trpc-users.service.ts
@Injectable()
export class TRCPUsersService {
  constructor(
    private readonly trpcInit: TRCPInitService,
  ) {}
  usersRouter = this.trpcInit.t.procedure.query(...)
} 
```

```ts
// trpc.service.ts
@Injectable()
export class TRCPService {
  constructor(
    private readonly trpcInit: TRCPInitService,
    private readonly trpcUsers: TRCPUsersService,
  ) {}
  appRouter = this.trpcInit.t.router({
    users: this.trpcUsers.usersRouter
  })
}
```

This exact approach is now also described in https://www.tomray.dev/nestjs-nextjs-trpc with more details and examples. 

For new projects I would suggest to take a look at https://github.com/ts-rest/ts-rest which has first-party support for NestJS.

**Reply by AlexSua** [reply]
That was not the problem. I was using "esnext" as target in typescript compiler options and it was not allowing me to do this call:
```js
 this.trpcInit
```
as it was detecting that it was a non initialized variable. I changed the target to "ES2022" and now it works. My bad.

I believed as well it was not possible to do this call:
```js
  appRouter = this.trpcInit.t.router({
    users: this.trpcUsers.usersRouter
  })
```
since in my head this part of the code was executed before the constructor.

**Reply by iceStorm** [reply]
I was recently also facing this problem. I'm specifying ESNext for the entire app. To fix the above issue, I ended up using this solution: https://github.com/microsoft/TypeScript/issues/53286#issuecomment-1500634952

TLDR;
Set `useDefineForClassFields: false` inside tsconfig of the NestJS app.

**Comment by Brokerful** [comment]
hi, how would you use it on the client side?

**Reply by reslear** [reply]
in monorepo **api** package

**Comment by macstr1k3r** [comment]
Hi,

I've built an adapter that will enable you to use tRPC with Nest.JS.

have a look at it, and if it gains some traction I'll work on documenting it properly.

https://github.com/macstr1k3r/trpc-nestjs-adapter

**Reply by yf-hk** [reply]
Looks cool!
I wonder where should I put code like below into a project?
![image](https://user-images.githubusercontent.com/203980/214979332-5608d24e-3d9e-4b47-b896-effe968c64f6.png)

**Reply by macstr1k3r** [reply]
It's kinda up to you.

Nest.JS's interpretation of controllers vastly differs from tRPC's.

I use some form of domain oriented folder structure, so in every domain folder I introduced 2 folders, `queries` and `mutations`.

All of those are gathered in a domain router which I later on connect to the root tRPC router.

The end result is roughly.
`trpcClient.domainRouter.query`

In the future I intend to export the zod schemas separately so that I can import them on the front-end and use them to validate forms, where applicable.

**Comment by samchon** [comment]
What about using [nestia](https://github.com/samchon/nestia) in NestJS?

It can generate SDK library and the SDK library can be same thing like `createTRPCProxyClient` of `trpc`.

**Comment by yousif-khalil-abdulkarim** [comment]
You coudl try use [typedi](https://docs.typestack.community/typedi/) standalone depdency injection library or only use nest js as [depedency injection library](https://docs.nestjs.com/standalone-applications)

**Comment by sannajammeh** [comment]
ts-rest was made specifically for allowing this type of task. Very similar to tRPC

**Reply by yf-hk** [reply]
ts-rest seems to be pretty cool. Thanks for sharing!

**Reply by firxworx** [reply]
@dzcpy also check out https://www.zodios.org/ - _typescript http client and server with zod validation_

**Reply by robert-king** [reply]
we're using ts-rest, pretty nice especially with NX

**Comment by cloudever** [comment]
If someone would like to write typed end-to-end microservices using nestjs, just use following approach, same implemented in trpc using js proxies

**Gateway side:**
```ts
import type { UsersService } from '@myproject/users-service' // <-- here we just import types, not all
import { InjectClientServiceProxy } from '../decorators'

@Resolver()
export class ExchangeDirectionResolver {
  // Please note that `USERS_SERVICE` - is just a token defined in ClientsModule
  @InjectClientServiceProxy({ name: "USERS_SERVICE" })
  private readonly usersService: UsersService

  @Query()
  async getUser(@Args() id: string) {
    return this.usersService.users.getById(id)
  }
}
```

And `@InjectClientServiceProxy` decorator that uses power of proxies
```ts
import { ClientProxy } from '@nestjs/microservices'
import { Inject, HttpException, HttpStatus } from '@nestjs/common'
import DeepProxy from 'proxy-deep'

export const InjectClientServiceProxy =
  (options: { name: string }): PropertyDecorator =>
  (target, property) => {
    const clientProperty = `${String(property)}__client`

    Inject(options.name)(target, clientProperty)

    let getInstance = () => {} // <-- that might be anti-pattern but I've not found any other solutions to class instance

    const proxy = new DeepProxy(
      {},
      {
        get() {
          return this.nest(function () {})
        },
        apply(this, applyTarget, thisArg, args) {
          const promise = (getInstance()[clientProperty] as ClientProxy)
            .send(this.path.join('.'), args[0])
            .toPromise()

          return promise
        },
      }
    )

    Object.defineProperty(target, property, {
      get() {
        getInstance = () => {
          return this
        }

        return proxy
      },
    })
  }
```

**Microservice side**

`UsersService` is a class that bootstap all data-access services in users microservice 
```ts
@Injectable()
@RegisterGlobalServiceContoller() // <-- does some magic
export class UsersService {
  @Inject(UserServiceDatabase)
  readonly users: UserServiceDatabase // <-- its a simple class with crud
  
  // here might be other data-access service just like `tokenService`, `resetPasswordService`
```

And `RegisterGlobalServiceContoller` decorator looks like:

```ts
export const RegisterGlobalServiceContoller =
  (): ClassDecorator => (Target) => {
    const metadata: { key: string; type: any }[] = Reflect.getMetadata(
      PROPERTY_DEPS_METADATA, // <-- imported from nestjs, thanks to nest runtime
      Target
    )

    metadata.forEach((meta) => {
      Controller()(meta.type) // <-- make data-access-service a controller else it cant be scanned by nestjs runtime

      Object.getOwnPropertyNames(meta.type.prototype).forEach((property) => {
        if (property === 'constructor') {
          return
        }

        MessagePattern(`${meta.key}.${property}`)(
          meta.type,
          property,
          Object.getOwnPropertyDescriptor(meta.type.prototype, property)
        )
      })
    })
  }
```
Don't forget that you must `export type UsersService` from `main.ts` and define all data-access-services in module `controllers: [...]`

**Troubles**
1. Do not use arrow style function in data-access service, else it cant be found by `RegisterGlobalServiceContoller`.
2. There must be only 1 argument in function calls between microservices

**Comment by Amuleus** [comment]
> I wonder how suitable this library is for Nest.js?

Surely possible. Please check a great tutorial from Tom Ray:
https://www.tomray.dev/nestjs-nextjs-trpc
https://github.com/tomwray13/nestjs-nextjs-trpc

**Comment by sannajammeh** [comment]
Just going to add that if you're not in need of Nest's DI system you can just bypass it all together by running your own instance of express or fastify and passing it into the Nest.js factory.

```tsx
	await fastify.register(fastifyTRPCPlugin as any, {
		prefix: "/trpc",
		trpcOptions: {
			router: appRouter,
			createContext: async (opts: CreateFastifyContextOptions) => {
				// Init payload
				return await trpcInnerContext({
					...opts,
					prisma,
					payload: initializedPayload,
					supabase,
					revenueCat,
				});
			},
		},
	});

	const nest = await NestFactory.create<NestFastifyApplication>(
		AppModule,
		new FastifyAdapter(fastify as any), // Type conflicts here
	);

```

**Comment by merthanmerter** [comment]
> None of the maintainers of tRPC are users of nestjs. 
> 
> 
> 
> Therefore, we have no plans on putting in more unpaid work to support nestjs when we have no use of it ourselves - this has to come from the community or through paid consulting.
> 
> 
> 
> I have no idea what it would take to support nest as I'm not a user of it so I have no insight on how an adapter would look like. 
> 
> 
> 
> ---
> 
> 
> 
> **Update:** [see potential workaround by @hypnodron below](https://github.com/trpc/trpc/discussions/1504#discussioncomment-4218180)

sorry but remove this section from your website then;

Framework agnostic
Compatible with all JavaScript frameworks and runtimes. It's easy to add to your existing projects.

**Comment by KevinEdry** [comment]
I've always felt that the workarounds that were suggested here were a good starting point, but there was nothing that actually takes into consideration the opinionated way of writing backend in nestjs.

So I've build this: https://www.nestjs-trpc.io/

The integration between the two it uses code generation to generate the trpc router types and uses both nestjs and trpc idioms with the a "Router" and "Procedure" concepts instead of controllers and services.

It works but not perfectly and I'm still working on the docs and readme, but with your feedback and help we can bring this to fruition.
